### PR TITLE
[PSM Interop] Print the hint for the triager on blanket errors

### PR DESCRIPTION
**Pull request by sergiitk**
This clearly indicates which errors are "blanket" errors and are not a root cause on their own.
This also moves the debug info with the last known status of an object the framework was waiting for, but bailed out due to a timeout. Previously it was printed as the last error message in the test, and this PR prints it after the stack trace that caused the test failure.
In addition, I added a similar debug information to the "wait for NEGs to become healthy". Now it prints the statuses of unhealthy backends

To achieve that, I mimicked upcoming  [PEP 678](https://peps.python.org/pep-0678/) Exception notes feature. When we're upgrade to py11, we'll be able to remove `add_note()` methods, and get the same functionality for free.

Screenshot 2023-08-07 at 3 05 59 PM

### CODE REVIEW

**Comment by sergiitk** [comment]
cc @gnossen @eugeneo

**Comment by sergiitk** [comment]
- [ ] [grpc/java/v1.57.x/branch/xds_k8s_lb](https://source.cloud.google.com/results/invocations/ad65f0b3-c9e6-4cdb-a749-cd472580ade2)
- [ ] [grpc/java/v1.57.x/branch/psm-security](https://source.cloud.google.com/results/invocations/8c540630-8a79-4dbb-b640-c84377238128)
- [ ] [grpc/java/v1.57.x/branch/xds_url_map](https://source.cloud.google.com/results/invocations/557ee800-4173-4fdd-9447-42959919e9f8)

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/gcp/compute.py [hunk 1]
```diff
@@ -392,14 +393,56 @@ def wait_for_backends_healthy_status(
...
+            timeout=timeout,
             check_result=lambda result: result,
         )
         pending = set(backends)
-        retryer(self._retry_backends_health, backend_service, pending)
+        try:
+            retryer(self._retry_backends_health, backend_service, pending)
+        except retryers.RetryError as retry_err:
+            unhealthy_backends: str
+            unhealthy_backends = ",".join([backend.name for backend in pending])
```
Comment (XuanWang-Amos) [inline tools/run_tests/xds_k8s_test_driver/framework/infrastructure/gcp/compute.py:411]: [Minor] Combine those two lines?

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/gcp/compute.py [hunk 2]
```diff
@@ -392,14 +393,56 @@ def wait_for_backends_healthy_status(
...
+
+            # Attempt to load backend health info for better debug info.
+            try:
+                unhealthy = []
+                for backend in pending:
+                    unhealthy.append(
+                        self.get_backend_service_backend_health(
```
Comment (XuanWang-Amos) [inline tools/run_tests/xds_k8s_test_driver/framework/infrastructure/gcp/compute.py:418]: Just curious, this will only return unhealthy backend?

**Review by XuanWang-Amos — COMMENTED** [review]
Can we have some example logs (Preferably before and after comparisons)?

File: tools/run_tests/xds_k8s_test_driver/framework/errors.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    def add_note(self, note: str):
+        self.note = note
+
+    def __str__(self):
+        return self.message if not self.note else f"{self.message}\n{self.note}"
+
+    @classmethod
+    def note_blanket_error(cls, reason: str) -> str:
+        return (
+            f"Important!!! This is a blanket error, just indicating {reason}."
```
Comment (gnossen) [inline tools/run_tests/xds_k8s_test_driver/framework/errors.py:42]: I think this needs to be more prominent. How about ALL_CAPS and a flower box around it?

File: tools/run_tests/xds_k8s_test_driver/framework/errors.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            f" Further investigation must be done to determine the root cause."
```
Comment (gnossen) [inline tools/run_tests/xds_k8s_test_driver/framework/errors.py:43]: Let's specifically mention that interop client and/or server logs should be inspected.

File: tools/run_tests/xds_k8s_test_driver/framework/errors.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    def __init__(self, message: str, *args, **kwargs):
+        self.message = message
+        # Exception only stores args.
+        self.kwargs = kwargs
+        # Pass to the Exception as if message is in **args.
+        super().__init__(*[message, *args])
+
+    # TODO(sergiitk): Remove in py3.11, this will be built-in. See PEP 678.
+    def add_note(self, note: str):
```
Comment (gnossen) [inline tools/run_tests/xds_k8s_test_driver/framework/errors.py:33]: Very fancy. :+1:

File: tools/run_tests/xds_k8s_test_driver/framework/errors.py (hunk 3 above)
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/errors.py:42]: How serious are you about this? 🌻🌻🌻

**Comment by sergiitk** [comment]
@XuanWang-Amos I ended up posting the screenshot. Unfortunately these are hard to reproduce to make capture all the before/after cases.

File: tools/run_tests/xds_k8s_test_driver/framework/errors.py (hunk 3 above)
Comment (gnossen) [reply tools/run_tests/xds_k8s_test_driver/framework/errors.py:42]: I meant more like:

```
##########################################################
# IMPORTANT: This is not a root cause. This is an indication
# _something_ -- literally _anything_ -- has gone wrong in the xDS
# flow. It is _your_ responsibility to look through the interop client and/or
# server logs to determine what exactly went wrong.
##########################################################
```

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/gcp/compute.py (hunk 2 above)
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/infrastructure/gcp/compute.py:418]: No. At this moment everything is `pending` is not healthy, but it's possible that between we gave up and printed this error the backend transitioned to healthy. I'll add a comment there.

File: tools/run_tests/xds_k8s_test_driver/framework/errors.py (hunk 3 above)
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/errors.py:42]: Done, here's the new format:
Screenshot 2023-08-07 at 3 05 59 PM

File: tools/run_tests/xds_k8s_test_driver/framework/errors.py (hunk 4 above)
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/errors.py]: Replaced the copy with the one you recommended below.

**Comment by sergiitk** [comment]
Just in case:
- [ ] [grpc/java/v1.57.x/branch/xds_k8s_lb](https://source.cloud.google.com/results/invocations/6568eefb-e6c6-4e52-ba15-297021515e28)
- [ ] [grpc/java/v1.57.x/branch/psm-security](https://source.cloud.google.com/results/invocations/4c44e97b-d5de-4666-bd2b-582edf777ea8)
- [ ] [grpc/java/v1.57.x/branch/xds_url_map](https://source.cloud.google.com/results/invocations/dd32cbe8-315f-4caa-b7af-ca2b1ad4a6c6)

**Pull request merged**